pvlib/singlediode_methods.py

```diff
@@ -0,0 +1,347 @@
+try:
+    from scipy.optimize import _array_newton
+except ImportError:
+    from pvlib import tools
+    from pvlib.tools import _array_newton
+# rename newton and set keyword arguments
+newton = partial(_array_newton, tol=1e-6, maxiter=100, fprime2=None)
+
+
+def estimate_voc(photocurrent, saturation_current, nNsVth):
+    """
+    Rough estimate of open circuit voltage useful for bounding searches for
+    ``i`` of ``v`` when using :func:`~pvlib.pvsystem.singlediode`.
+
+    Parameters
+    ----------
+    photocurrent : numeric
+        photo-generated current [A]
+    saturation_current : numeric
+        diode reverse saturation current [A]
+    nNsVth : numeric
+        product of thermal voltage ``Vth`` [V], diode ideality factor ``n``,
+        and number of series cells ``Ns``
+
+    Returns
+    -------
+    numeric
+        rough estimate of open circuit voltage [V]
+
+    Notes
+    -----
+    Calculating the open circuit voltage, :math:`V_{oc}`, of an ideal device
+    with infinite shunt resistance, :math:`R_{sh} \\to \\infty`, and zero series
+    resistance, :math:`R_s = 0`, yields the following equation [1]. As an
+    estimate of :math:`V_{oc}` it is useful as an upper bound for the bisection
+    method.
+
+    .. math::
+
+        V_{oc, est}=n Ns V_{th} \\log \\left( \\frac{I_L}{I_0} + 1 \\right)
+
+    [1] http://www.pveducation.org/pvcdrom/open-circuit-voltage
+    """
+
+    return nNsVth * np.log(np.asarray(photocurrent) / saturation_current + 1.0)
+
+
+def bishop88(diode_voltage, photocurrent, saturation_current, resistance_series,
+             resistance_shunt, nNsVth, gradients=False):
+    """
+    Explicit calculation of points on the IV curve described by the single
+    diode equation [1].
+
+    [1] "Computer simulation of the effects of electrical mismatches in
+    photovoltaic cell interconnection circuits" JW Bishop, Solar Cell (1988)
+    https://doi.org/10.1016/0379-6787(88)90059-2
+
+    Parameters
+    ----------
+    diode_voltage : numeric
+        diode voltages [V]
+    photocurrent : numeric
+        photo-generated current [A]
+    saturation_current : numeric
+        diode reverse saturation current [A]
+    resistance_series : numeric
+        series resistance [ohms]
+    resistance_shunt: numeric
+        shunt resistance [ohms]
+    nNsVth : numeric
+        product of thermal voltage ``Vth`` [V], diode ideality factor ``n``,
+        and number of series cells ``Ns``
+    gradients : bool
+        False returns only I, V, and P. True also returns gradients
+
+    Returns
+    -------
+    tuple
+        currents [A], voltages [V], power [W], and optionally
+        :math:`\\frac{dI}{dV_d}`, :math:`\\frac{dV}{dV_d}`,
+        :math:`\\frac{dI}{dV}`, :math:`\\frac{dP}{dV}`, and
+        :math:`\\frac{d^2 P}{dV dV_d}`
+    """
+    a = np.exp(diode_voltage / nNsVth)
+    b = 1.0 / resistance_shunt
+    i = photocurrent - saturation_current * (a - 1.0) - diode_voltage * b
```

Comment:
Perhaps [np.expm1](https://docs.scipy.org/doc/numpy/reference/generated/numpy.expm1.html) is preferable here? If you do change it here, I guess I'm not sure what to do about the ``a`` term below. I guess I'd put the ``exp`` call in line below.

Reply:
IMHO no, `np.expm1` is for very small values of `x`, so this might benefit very small voltages close to V=0 (at short circuit), but IMO errors there are not as important? But I am open to considering it if others are in favor. Is there an additional computational expense to using `np.expm1`? Can we save this as another PR?

for typical range of voltage near Vmp there's no difference between `np.expm1` and `np.exp`, but as V &rarr; zero, the error increases, slightly (less than 1 ppm)
```python
In [10]: np.expm1(0.5/1.23/0.026)
Out[10]: 6167299.206747167

In [11]: np.exp(0.5/1.23/0.026) - 1.0
Out[11]: 6167299.206747167

In [12]: np.expm1(1e-10)
Out[12]: 1.00000000005e-10

In [13]: np.exp(1e-10) - 1.0
Out[13]: 1.000000082740371e-10
```

Reply:
Won't this function be called with a range of voltages from 0 to v_oc? I agree that for most applications this doesn't matter at all. But some contributors have been more interested in precision and "gold" datasets/algorithms lately. Maybe a couple more digits of precision avoids some issue in the future with very little additional cost. Or maybe it's waste. I don't know.

Reply:
But we tested the numerical precision using SymPy and there were no differences within the tolerance [(the defaults are `rtol=1e-5` and `atol=1e-8`)](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html). Anyway, I added an issue, #500, for it and we can consider it altogether.

Reply:
✔️ close #500